// mf-react-csv-to-table/dist/index.js
'use strict';

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var React__default = _interopDefault(React);
var PropTypes = _interopDefault(require('prop-types'));

function createCommonjsModule(fn, module) {
	return module = { exports: {} }, fn(module, module.exports), module.exports;
}

var classnames = createCommonjsModule(function (module) {
/*!
  Copyright (c) 2017 Jed Watson.
  Licensed under the MIT License (MIT), see
  http://jedwatson.github.io/classnames
*/
/* global define */

(function () {

	var hasOwn = {}.hasOwnProperty;

	function classNames () {
		var classes = [];

		for (var i = 0; i < arguments.length; i++) {
			var arg = arguments[i];
			if (!arg) continue;

			var argType = typeof arg;

			if (argType === 'string' || argType === 'number') {
				classes.push(arg);
			} else if (Array.isArray(arg) && arg.length) {
				var inner = classNames.apply(null, arg);
				if (inner) {
					classes.push(inner);
				}
			} else if (argType === 'object') {
				for (var key in arg) {
					if (hasOwn.call(arg, key) && arg[key]) {
						classes.push(key);
					}
				}
			}
		}

		return classes.join(' ');
	}

	if (module.exports) {
		classNames.default = classNames;
		module.exports = classNames;
	} else {
		window.classNames = classNames;
	}
}());
});

function styleInject(css, ref) {
  if ( ref === void 0 ) ref = {};
  var insertAt = ref.insertAt;

  if (!css || typeof document === 'undefined') { return; }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css = "/* add css styles here (optional) */\n\n.styles_nav_pagination__3Lo0d {\n    display: inline-block;\n}\n\n.styles_nav_pagination__3Lo0d > ul {\n    list-style: none;\n    padding: 0 15px;\n    display: flex;\n    flex-wrap: wrap;\n    justify-content: center;\n    align-items: center;\n}\n\n.styles_nav_pagination__3Lo0d > ul > li {\n    padding: 0 5px;\n}\n\n.styles_nav_pagination__3Lo0d > ul > li > a {\n    text-decoration: none;\n    color: inherit;\n}";
var styles = { "nav_pagination": "styles_nav_pagination__3Lo0d" };
styleInject(css);

var classCallCheck = function (instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
};

var createClass = function () {
  function defineProperties(target, props) {
    for (var i = 0; i < props.length; i++) {
      var descriptor = props[i];
      descriptor.enumerable = descriptor.enumerable || false;
      descriptor.configurable = true;
      if ("value" in descriptor) descriptor.writable = true;
      Object.defineProperty(target, descriptor.key, descriptor);
    }
  }

  return function (Constructor, protoProps, staticProps) {
    if (protoProps) defineProperties(Constructor.prototype, protoProps);
    if (staticProps) defineProperties(Constructor, staticProps);
    return Constructor;
  };
}();

var defineProperty = function (obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }

  return obj;
};

var inherits = function (subClass, superClass) {
  if (typeof superClass !== "function" && superClass !== null) {
    throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
  }

  subClass.prototype = Object.create(superClass && superClass.prototype, {
    constructor: {
      value: subClass,
      enumerable: false,
      writable: true,
      configurable: true
    }
  });
  if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
};

var possibleConstructorReturn = function (self, call) {
  if (!self) {
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  }

  return call && (typeof call === "object" || typeof call === "function") ? call : self;
};

var Pagination = function (_Component) {
    inherits(Pagination, _Component);

    function Pagination(props) {
        classCallCheck(this, Pagination);

        var _this = possibleConstructorReturn(this, (Pagination.__proto__ || Object.getPrototypeOf(Pagination)).call(this, props));

        _this.state = {
            renderPageNumbers: ''
        };
        return _this;
    }

    createClass(Pagination, [{
        key: 'componentDidMount',
        value: function componentDidMount() {
            this.buildPages();
        }
    }, {
        key: 'prev',
        value: function prev(e) {
            var page = this.props.options.current > 1 ? this.props.options.current - 1 : 1;
            this.toPage(e, page);
        }
    }, {
        key: 'next',
        value: function next(e) {
            var page = this.props.options.current + 1;
            this.toPage(e, page);
        }
    }, {
        key: 'toPage',
        value: function toPage(e, page) {
            e.preventDefault();
            if (page !== this.props.options.current && !(page > this.props.options.last || page < 1)) {
                this.props.onClick(page);
                this.props.options.current = page;

                this.buildPages();
            }
        }
    }, {
        key: 'buildPages',
        value: function buildPages() {
            var _this2 = this;

            var pageNumbers = [];
            var renderPageNumbers = void 0;

            var _props = this.props,
                pageRangeDisplayed = _props.pageRangeDisplayed,
                itemClass = _props.itemClass,
                linkClass = _props.linkClass,
                activeClass = _props.activeClass,
                disabledClass = _props.disabledClass;


            if (this.props.options.total > 0) {

                for (var i = 1; i <= Math.ceil(this.props.options.total / this.props.options.perPage); i++) {
                    pageNumbers.push(i);
                }

                renderPageNumbers = pageNumbers.filter(function (number) {
                    return number == 1 && number >= _this2.props.options.current - pageRangeDisplayed || number == _this2.props.options.total || number >= _this2.props.options.current - pageRangeDisplayed && number <= _this2.props.options.current + pageRangeDisplayed;
                }).map(function (number) {
                    var _cx;

                    var classes = classnames(itemClass, (_cx = {}, defineProperty(_cx, activeClass, _this2.props.options.current === number), defineProperty(_cx, disabledClass, false), _cx));
                    return React__default.createElement(
                        'li',
                        { key: number, className: classes },
                        React__default.createElement(
                            'a',
                            { className: linkClass, href: '#', onClick: function onClick(e) {
                                    return _this2.toPage(e, number);
                                }, 'aria-label': 'Go to page number ' + number },
                            number
                        )
                    );
                });
            }

            this.setState({ renderPageNumbers: renderPageNumbers });
        }
    }, {
        key: 'render',
        value: function render() {
            var _this3 = this;

            var renderPageNumbers = this.state.renderPageNumbers;
            var _props2 = this.props,
                innerClass = _props2.innerClass,
                itemClass = _props2.itemClass,
                linkClass = _props2.linkClass,
                disabledClass = _props2.disabledClass;


            return React__default.createElement(
                'nav',
                { className: styles.nav_pagination },
                React__default.createElement(
                    'ul',
                    { className: innerClass },
                    React__default.createElement(
                        'li',
                        { className: classnames(itemClass, defineProperty({}, disabledClass, this.props.options.current == 1))
                        },
                        React__default.createElement(
                            'a',
                            { className: linkClass, href: '#', onClick: function onClick(e) {
                                    return _this3.toPage(e, 1);
                                } },
                            this.props.firstPageText
                        )
                    ),
                    React__default.createElement(
                        'li',
                        { className: classnames(itemClass, defineProperty({}, disabledClass, this.props.options.current === 1))
                        },
                        React__default.createElement(
                            'a',
                            { className: linkClass, href: '#', onClick: function onClick(e) {
                                    return _this3.prev(e);
                                } },
                            this.props.prevPageText
                        )
                    ),
                    renderPageNumbers,
                    React__default.createElement(
                        'li',
                        { className: classnames(itemClass, defineProperty({}, disabledClass, this.props.options.current === this.props.options.last))
                        },
                        React__default.createElement(
                            'a',
                            { className: linkClass, href: '#', onClick: function onClick(e) {
                                    return _this3.next(e);
                                } },
                            this.props.nextPageText
                        )
                    ),
                    React__default.createElement(
                        'li',
                        { className: classnames(itemClass, defineProperty({}, disabledClass, this.props.options.last == this.props.options.current))
                        },
                        React__default.createElement(
                            'a',
                            { className: linkClass, href: '#', onClick: function onClick(e) {
                                    return _this3.toPage(e, _this3.props.options.last);
                                } },
                            this.props.lastPageText
                        )
                    )
                )
            );
        }
    }]);
    return Pagination;
}(React.Component);

Pagination.propTypes = {
    onClick: PropTypes.func.isRequired,
    options: PropTypes.object.isRequired,
    pageRangeDisplayed: PropTypes.number,
    prevPageText: PropTypes.string,
    firstPageText: PropTypes.string,
    nextPageText: PropTypes.string,
    lastPageText: PropTypes.string,
    innerClass: PropTypes.string,
    itemClass: PropTypes.string,
    linkClass: PropTypes.string,
    activeClass: PropTypes.string,
    disabledClass: PropTypes.string
};
Pagination.defaultProps = {
    pageRangeDisplayed: 3,
    prevPageText: "⟨",
    firstPageText: "«",
    nextPageText: "⟩",
    lastPageText: "»",
    innerClass: "pagination",
    itemClass: undefined,
    linkClass: undefined,
    activeClass: "active",
    disabledClass: "disabled"
};

function styleInject$1(css, ref) {
  if ( ref === void 0 ) ref = {};
  var insertAt = ref.insertAt;

  if (!css || typeof document === 'undefined') { return; }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css$1 = "/* add css styles here (optional) */\n.styles_paginator__1voR3 {\n  text-align: center;\n  display: block;\n}";
var styles$1 = { "paginator": "styles_paginator__1voR3" };
styleInject$1(css$1);

var classCallCheck$1 = function (instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
};

var createClass$1 = function () {
  function defineProperties(target, props) {
    for (var i = 0; i < props.length; i++) {
      var descriptor = props[i];
      descriptor.enumerable = descriptor.enumerable || false;
      descriptor.configurable = true;
      if ("value" in descriptor) descriptor.writable = true;
      Object.defineProperty(target, descriptor.key, descriptor);
    }
  }

  return function (Constructor, protoProps, staticProps) {
    if (protoProps) defineProperties(Constructor.prototype, protoProps);
    if (staticProps) defineProperties(Constructor, staticProps);
    return Constructor;
  };
}();

var inherits$1 = function (subClass, superClass) {
  if (typeof superClass !== "function" && superClass !== null) {
    throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
  }

  subClass.prototype = Object.create(superClass && superClass.prototype, {
    constructor: {
      value: subClass,
      enumerable: false,
      writable: true,
      configurable: true
    }
  });
  if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
};

var possibleConstructorReturn$1 = function (self, call) {
  if (!self) {
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  }

  return call && (typeof call === "object" || typeof call === "function") ? call : self;
};

var CsvToTable = function (_Component) {
  inherits$1(CsvToTable, _Component);

  function CsvToTable(props) {
    classCallCheck$1(this, CsvToTable);

    var _this = possibleConstructorReturn$1(this, (CsvToTable.__proto__ || Object.getPrototypeOf(CsvToTable)).call(this, props));

    _this.state = {
      displayCsvData: [],
      csvHead: null,
      csvData: null,
      pagination: null
    };
    return _this;
  }

  createClass$1(CsvToTable, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.handleFile();
    }
  }, {
    key: 'handleFile',
    value: function handleFile() {
      var _this2 = this;

      var csv = this.props.csv;

      var reader = new FileReader();
      reader.onload = function () {
        _this2.parseCsvData(reader.result);
      };
      reader.readAsText(csv);
    }
  }, {
    key: 'parseCsvData',
    value: function parseCsvData(data) {
      var _this3 = this;

      var dataFomatted = data.replace(/\s*[\r\n]"/gm, '"\n').replace(/^\s*[\r\n]/gm, '');

      var rows = dataFomatted.split('\n');
      var csvData = [];

      rows.forEach(function (row) {
        csvData.push(row.split(_this3.props.separator));
      });

      this.setState({
        csvHead: this.props.hasHead ? csvData[0] : null,
        csvData: this.props.hasHead ? csvData.slice(1) : csvData.slice()
      });

      setTimeout(function () {
        _this3.showCsvData(1);
      }, 500);
    }
  }, {
    key: 'showCsvData',
    value: function showCsvData(page) {
      if (this.state.csvData) {
        var total = this.state.csvData.length;
        var last = total / this.props.perPage;
        var totalByPage = page * this.props.perPage;

        if (page >= 1 && page <= last) {
          var start = totalByPage - this.props.perPage;
          var max = totalByPage > total ? start + (totalByPage - total) : totalByPage;
          var displayCsvData = [];

          for (var index = start; index < max; index++) {
            displayCsvData.push(this.state.csvData[index]);
          }

          this.setState({
            pagination: {
              perPage: this.props.perPage,
              current: page,
              prev: page > 1,
              next: total > 1 && page < last,
              last: last,
              total: total
            },
            displayCsvData: displayCsvData
          });
        }
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          title = _props.title,
          pageRangeDisplayed = _props.pageRangeDisplayed,
          prevPageText = _props.prevPageText,
          firstPageText = _props.firstPageText,
          nextPageText = _props.nextPageText,
          lastPageText = _props.lastPageText,
          innerClass = _props.innerClass,
          paginatorInnerClass = _props.paginatorInnerClass,
          itemClass = _props.itemClass,
          linkClass = _props.linkClass,
          activeClass = _props.activeClass,
          disabledClass = _props.disabledClass;


      return React__default.createElement(
        React__default.Fragment,
        null,
        this.props.csv && React__default.createElement(
          'div',
          { className: 'mf-react-csv-to-table ' + (innerClass || '') },
          title && React__default.createElement(
            'h3',
            null,
            title
          ),
          React__default.createElement(
            'div',
            { className: 'table-responsive' },
            React__default.createElement(
              'table',
              { className: 'table' },
              this.props.hasHead && this.state.csvHead && React__default.createElement(
                'thead',
                null,
                React__default.createElement(
                  'tr',
                  null,
                  this.state.csvHead.map(function (head, h) {
                    return React__default.createElement(
                      'th',
                      { key: 'head-' + h },
                      head
                    );
                  })
                )
              ),
              React__default.createElement(
                'tbody',
                null,
                this.state.displayCsvData.map(function (row, i) {
                  return React__default.createElement(
                    'tr',
                    { key: 'row-' + i },
                    row.map(function (col, j) {
                      return React__default.createElement(
                        'td',
                        { key: 'col-' + j },
                        col
                      );
                    })
                  );
                })
              )
            ),
            React__default.createElement(
              'div',
              { className: styles$1.paginator },
              this.state.pagination && React__default.createElement(Pagination, {
                onClick: this.showCsvData.bind(this),
                options: this.state.pagination,
                pageRangeDisplayed: pageRangeDisplayed,
                prevPageText: prevPageText,
                firstPageText: firstPageText,
                nextPageText: nextPageText,
                lastPageText: lastPageText,
                innerClass: paginatorInnerClass,
                itemClass: itemClass,
                linkClass: linkClass,
                activeClass: activeClass,
                disabledClass: disabledClass
              })
            )
          )
        )
      );
    }
  }]);
  return CsvToTable;
}(React.Component);

CsvToTable.propTypes = {
  csv: PropTypes.object,
  separator: PropTypes.string,
  perPage: PropTypes.number,
  hasHead: PropTypes.bool,
  pageRangeDisplayed: PropTypes.number,
  prevPageText: PropTypes.string,
  firstPageText: PropTypes.string,
  nextPageText: PropTypes.string,
  lastPageText: PropTypes.string,
  innerClass: PropTypes.string,
  paginatorInnerClass: PropTypes.string,
  itemClass: PropTypes.string,
  linkClass: PropTypes.string,
  activeClass: PropTypes.string,
  disabledClass: PropTypes.string
};
CsvToTable.defaultProps = {
  separator: ',',
  perPage: 15,
  hasHead: true
};

module.exports = CsvToTable;
